src/main/java/reactor/kafka/receiver/internals/DefaultKafkaReceiver.java

```diff
@@ -25,135 +24,141 @@
 
 import reactor.core.publisher.Flux;
 import reactor.core.publisher.Mono;
-import reactor.core.scheduler.Scheduler;
-import reactor.core.scheduler.Schedulers;
 import reactor.kafka.receiver.ReceiverOptions;
 import reactor.kafka.receiver.ReceiverRecord;
 import reactor.kafka.receiver.KafkaReceiver;
 import reactor.kafka.sender.TransactionManager;
-import reactor.util.context.Context;
 
 public class DefaultKafkaReceiver<K, V> implements KafkaReceiver<K, V> {
 
-    static final String AWAITING_TRANSACTION_KEY = "reactor.kafka.receiver.internals.awaitingTransaction";
-
     private final ConsumerFactory consumerFactory;
 
     private final ReceiverOptions<K, V> receiverOptions;
 
-    private Scheduler scheduler;
-
-    ConsumerFlux<K, V> consumerFlux;
-
     Predicate<Throwable> isRetriableException = RetriableCommitFailedException.class::isInstance;
 
+    ConsumerHandler<K, V> consumerHandler;
+
     public DefaultKafkaReceiver(ConsumerFactory consumerFactory, ReceiverOptions<K, V> receiverOptions) {
         this.consumerFactory = consumerFactory;
         this.receiverOptions = receiverOptions.toImmutable();
     }
 
+    private Mono<ConsumerHandler<K, V>> start(AckMode ackMode) {
+        return Mono.fromCallable(() -> {
+            return consumerHandler = new ConsumerHandler<>(
+                receiverOptions,
+                consumerFactory.createConsumer(receiverOptions),
+                // Always use the currently set value
+                e -> isRetriableException.test(e),
+                ackMode
+            );
+        });
+    }
+
     @Override
     public Flux<ReceiverRecord<K, V>> receive() {
-        ConsumerFlux<K, V> consumerFlux = createConsumerFlux(AckMode.MANUAL_ACK);
-        return consumerFlux
-            .onBackpressureBuffer()
-            .publishOn(scheduler)
-            .doFinally(signal -> dispose())
-            .flatMapIterable(it -> it)
-            .map(record -> new ReceiverRecord<>(
-                record,
-                consumerFlux.new CommittableOffset(record)
-            ))
-            .doOnRequest(consumerFlux::handleRequest);
+        return Flux.usingWhen(
+            start(AckMode.MANUAL_ACK),
+            handler -> {
+                return handler
+                    .receive()
+                    .flatMapIterable(it -> it)
+                    .map(record -> new ReceiverRecord<>(
+                        record,
+                        handler.toCommittableOffset(record)
+                    ))
+                    .doOnRequest(handler::handleRequest);
+            },
+            this::dispose
```

Comment:
minor nitpick but I would just consider renaming the `dispose(ConsumerHandler)` method at the bottom to something along the lines of `disposeConsumerHandler` for clarity (when first reading the code my first thought was that there was a `Disposable` involved. granted, the ambiguity is not very strong...)

Reply:
👍 changed to `cleanup`